pandas/core/algorithms.py

```diff
@@ -168,7 +168,8 @@ def _ensure_data(
     elif is_categorical_dtype(values.dtype) and (
         is_categorical_dtype(dtype) or dtype is None
     ):
-        values = values.codes
+        # error: "ExtensionArray" has no attribute "codes"
+        values = values.codes  # type: ignore[attr-defined]
```

Comment:
because we have `is_categorical_dtype(values.dtype)` check, ok to use cast on `values` instead of ignore.